backend/capellacollab/projects/toolmodels/modelsources/git/interface.py

```diff
@@ -0,0 +1,144 @@
+) -> settings_git_models.DatabaseGitInstance:
+    """Get the corresponding git instance for a git model
+    The git instance is selected via the longest common prefix match.
+    """
+
+    instances_sorted_by_len = sorted(
+        settings_git_crud.get_git_instances(db),
+        key=lambda instance: len(instance.url),
+        reverse=True,
+    )
+
+    for instance in instances_sorted_by_len:
+        if git_model.path.startswith(instance.url):
+            return instance
+    raise fastapi.HTTPException(
+        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+        detail={
+            "reason": (
+                "No matching git instance was found for the primary git model.",
+                "Please contact your administrator.",
+            ),
+        },
+    )
+
+
+def git_instance_is_gitlab(
+    git_instance: settings_git_models.DatabaseGitInstance,
+) -> bool:
+    return git_instance.type == settings_git_models.GitType.GITLAB
+
+
+def git_instance_is_github(
+    git_instance: settings_git_models.DatabaseGitInstance,
+) -> bool:
+    return git_instance.type == settings_git_models.GitType.GITHUB
+
+
+def check_git_instance_has_api_url(
+    git_instance: settings_git_models.DatabaseGitInstance,
+):
+    if not git_instance.api_url:
+        raise fastapi.HTTPException(
+            status_code=fastapi.status.HTTP_500_INTERNAL_SERVER_ERROR,
+            detail={
+                "err_code": "GIT_INSTANCE_NO_API_ENDPOINT_DEFINED",
+                "reason": (
+                    "The used Git instance has no API endpoint defined.",
+                    "Please contact your administrator.",
+                ),
+            },
+        )
+
+
+def get_last_job_run_id_for_git_model(
+    db: orm.Session, job_name: str, git_model: git_models.DatabaseGitModel
+) -> tuple[settings_git_models.DatabaseGitInstance, str, tuple[str, str]]:
+    git_instance = get_git_instance_for_git_model(db, git_model)
+    return get_git_interface(git_instance).get_last_job_run_id_for_git_model(
+        job_name, git_model, git_instance
+    )
+
+
+def get_artifact_from_job_as_json(
+    project_id: str,
+    job_id: str,
+    trusted_path_to_artifact: str,
+    git_model: git_models.DatabaseGitModel,
+    git_instance: settings_git_models.DatabaseGitInstance,
+) -> dict:
+    return get_git_interface(git_instance).get_artifact_from_job_as_json(
+        project_id,
+        job_id,
+        trusted_path_to_artifact,
+        git_model,
+        git_instance,
+    )
+
+
+def get_artifact_from_job_as_content(
+    project_id: str,
+    job_id: str,
+    trusted_path_to_artifact: str,
+    git_model: git_models.DatabaseGitModel,
+    git_instance: settings_git_models.DatabaseGitInstance,
+) -> bytes:
+    return get_git_interface(git_instance).get_artifact_from_job_as_content(
+        project_id,
+        job_id,
+        trusted_path_to_artifact,
+        git_model,
+        git_instance,
+    )
```

Comment:
I don't really like that we have to implement each function on the interface and have to derive each time, which interface it is.

Instead, I would suggest the following: 
- Create a superclass / interface. In the interface, you can implement all functions which are independent from the specific instance. Public instance-specific methods should be implemented as abstractmethods.
- Use the factory pattern to determine the right subclass: https://en.wikipedia.org/wiki/Factory_method_pattern
- Each instance has specific sub-class. The class inherits from the subclass / interface and implements the abstract functions + optionally private functions.

Reply:
how would you name the class then? when referencing it in the routes.py files I named it `interface_class` but I don't think this is really a good option - do you have a good idea?